services/device/geolocation/geoclue_location_provider.cc

```diff
@@ -0,0 +1,377 @@
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/services/device/geolocation/geoclue_location_provider.h"
+
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "base/barrier_callback.h"
+#include "base/feature_list.h"
+#include "base/functional/bind.h"
+#include "base/functional/callback_forward.h"
+#include "base/functional/callback_helpers.h"
+#include "base/location.h"
+#include "base/logging.h"
+#include "base/memory/weak_ptr.h"
+#include "base/ranges/algorithm.h"
+#include "base/time/time.h"
+#include "components/dbus/thread_linux/dbus_thread_linux.h"
+#include "dbus/property.h"
+#include "services/device/public/cpp/device_features.h"
+#include "services/device/public/cpp/geolocation/geoposition.h"
+
+namespace device {
+
+namespace {
+
+constexpr char kServiceName[] = "org.freedesktop.GeoClue2";
+constexpr char kLocationInterfaceName[] = "org.freedesktop.GeoClue2.Location";
+constexpr char kClientInterfaceName[] = "org.freedesktop.GeoClue2.Client";
+constexpr char kManagerInterfaceName[] = "org.freedesktop.GeoClue2.Manager";
+constexpr char kManagerObjectPath[] = "/org/freedesktop/GeoClue2/Manager";
+constexpr char kBraveDesktopId[] = "com.brave.Browser";
+
+// This corresponds to the GeoClue2 AccuracyLevel: Neighborhood
+constexpr uint32_t kGeoClueAccuracyLow = 5;
+
+// This corresponds to the GeoClue2 AccuracyLevel: Exact
+constexpr uint32_t kGeoClueAccuracyHigh = 8;
+
+mojom::Geoposition GetErrorPosition() {
+  mojom::Geoposition response;
+  response.error_code = mojom::Geoposition_ErrorCode::POSITION_UNAVAILABLE;
+  response.error_message = "Unable to create instance of Geolocation API";
+  return response;
+}
+
+// Note: This method blocks because the call to NewSystemProvider is not
+// asynchronous, but it happens on the background geolocation thread.
+//
+// The easiest way to determine if a DBus service exists is to try and call a
+// method on it, and see if it fails. For this, I chose
+// |GeoClue2.Manager.GetClient| (this is cached on the GeoClue2 side, so it will
+// be the same client we get when we start our service).
+bool GeoClueAvailable() {
+  dbus::Bus::Options options;
+  options.bus_type = dbus::Bus::SYSTEM;
+  options.connection_type = dbus::Bus::PRIVATE;
+  auto bus = base::MakeRefCounted<dbus::Bus>(options);
+
+  dbus::ObjectProxy* proxy =
+      bus->GetObjectProxy(kServiceName, dbus::ObjectPath(kManagerObjectPath));
+
+  dbus::MethodCall call(kManagerInterfaceName, "GetClient");
+  auto response =
+      proxy->CallMethodAndBlock(&call, dbus::ObjectProxy::TIMEOUT_USE_DEFAULT);
+
+  // If the response is |nullptr| then the GeoClue2.Manager does not exist.
+  bool available = response.get();
+
+  // Shutdown this bus - we'll use one on the DBus thread for our actual
+  // provider.
+  bus->ShutdownAndBlock();
```

Comment:
why are you destroying the bus instead of passing it into `GeoClueLocationProvider`?

you should be able to create it with `dbus_thread_linux::GetTaskRunner()` right here and keep using it down the road.

Reply:
If I create the `Bus` with `dbus_thread_linux::GetTaskRunner` then I can't call blocking functions from this thread. Similarly, if I create the `Bus` in the `GeoClueLocationProvider` without setting `dbus_task_runner` some async functions don't work.

I was thinking of refactoring the `GeoLocationProvider` to just use blocking `dbus` calls but @bridiver [suggested it might be better](https://github.com/brave/brave-core/pull/18063/files#r1195460755) to keep using the async calls.

Reply:
@fallaciousreasoning you cannot ever call blocking functions on the UI thread so this is not an option.

Reply:
@bridiver [services/device/geolocation/geoclue_location_provider.cc](https://github.com/brave/brave-core/pull/18063/files/82e67b75dafbe612f45d27d15fcebe8a0abe218e#diff-f9043cf9c7fbc9f2849c6a969d8b633b74b420126fafa8a03ed01cc279974e0b) runs on the UI thread.